src/libexpr/nixexpr.hh

```diff
@@ -24,31 +27,113 @@ MakeError(RestrictedPathError, Error);
 struct Pos
 {
     Symbol file;
+    FileOrigin origin;
     uint32_t line;
-    FileOrigin origin:2;
-    uint32_t column:30;
-    Pos() : line(0), origin(foString), column(0) { };
-    Pos(FileOrigin origin, const Symbol & file, uint32_t line, uint32_t column)
-        : file(file), line(line), origin(origin), column(column) { };
-    operator bool() const
+    uint32_t column;
+
+    explicit operator bool() const { return line > 0; }
+};
+
+class PosIdx {
+    friend class PosTable;
+
+private:
+    uint32_t id;
+
+    explicit PosIdx(uint32_t id): id(id) {}
+
+public:
+    PosIdx() : id(0) {}
+
+    explicit operator bool() const { return id > 0; }
+
+    bool operator<(const PosIdx other) const { return id < other.id; }
+};
+
+class PosTable
+{
+public:
+    class Origin {
+        friend PosTable;
+    private:
+        // must always be invalid by default, add() replaces this with the actual value.
+        // subsequent add() calls use this index as a token to quickly check whether the
+        // current origins.back() can be reused or not.
+        mutable uint32_t idx = std::numeric_limits<uint32_t>::max();
```

Comment:
no, instances of this type double as origin information *within* the symbol table and as cookie values for *users* of the symbol table to enable reuse of origin entries. the idea is to have each site that creates positions allocate an `Origin` with an invalid `idx`, which the symbol table fills with a valid value when a position is created. each subsequent position creation need only check that the `idx` of the cookie is the `idx` of the last origin in the table and reuse that origin if it is, otherwise create a new origin and change the `idx` of the cookie. it turns out doing it this way is a lot more efficient than having each position creation pass the full file/origin pair and check those.

Reply:
Oh yes, didn’t think about nesting the parsing (dunno whether that happens in practice, but it’s definitely safer to assume that it can)